exercises/brandon_hewer/ex02_oo_basics/src/ShapeSorter.cpp

```diff
@@ -0,0 +1,187 @@
+// Includes
+//----------------------------------------------------------------------
+#include "ShapeSorter.h"
+
+/*
+ * Constructs a new ShapeSorter for the specified vector of shapes.
+ *
+ * @param shapes	A vector containing pointers to shape objects.
+ */
+ShapeSorter::ShapeSorter(const std::vector<Shape*>& shapes) {
+	m_shapes = shapes;
+}
+
+/*
+ * Destructor for this ShapeSorter.
+ */
+ShapeSorter::~ShapeSorter() {
+}
+
+/*
+ * Sorts the vector of shapes by area, in the specified sort direction.
+ *
+ * @param direction	The direction in which to sort:
+ *                  ShapeSorter::ASCENDING (true) - Sort by increasing area.
+ *                  ShapeSorter::DESCENDING (false) - Sort by decreasing area.
+ */
+ShapeSorter ShapeSorter::sortByArea(bool direction) const {
+	// Copy vector of shapes.
+	std::vector<Shape*> n_shapes(m_shapes);
+	bool(*sorter) (Shape*, Shape*);
+
+	// Check whether the specified direction is ascending or
+	// descending.
+	if (direction == ASCENDING) {
+
+		// Create a sorter function for ascending sorting
+		// by area.
+		sorter = [](Shape* s1, Shape* s2) -> bool {
+			return s1->area() < s2->area();
+		};
+	}
+	else {
+
+		// Create a sorter function for descending sorting
+		// by area.
+		sorter = [](Shape* s1, Shape* s2) -> bool {
+			return s1->area() > s2->area();
+		};
+	}
+
+	// Sort and return new ShapeSorter for sorted vector
+	// of shapes.
+	sort(n_shapes.begin(), n_shapes.end(), sorter);
+	return ShapeSorter(n_shapes);
+}
+
+/*
+ * Sorts the vector of shapes by perimeter, in the specified sort direction.
+ * 
+ * @param direction	The direction in which to sort:
+ *					ShapeSorter::ASCENDING (true) - Sort by increasing perimeter.
+ *					ShapeSorter::DESCENDING (false) - Sort b
+ *
+ * @return			A ShapeSorter for the sorted vector of shapes.
+ */
+ShapeSorter ShapeSorter::sortByPerimeter(bool direction) const {
+	// Copy vector of shapes.
+	std::vector<Shape*> n_shapes(m_shapes);
+	bool(*sorter) (Shape*, Shape*);
+
+	// Check whether the specified direction is ascending or
+	// descending.
+	if (direction == ASCENDING) {
+
+		// Create a sorter function for ascending sorting
+		// by perimeter.
+		sorter = [](Shape* s1, Shape* s2) -> bool {
+			return s1->perimeter() < s2->perimeter();
+		};
+	}
+	else {
+
+		// Create a sorter function for descending sorting
+		// by perimeter.
+		sorter = [](Shape* s1, Shape* s2) -> bool {
+			return s1->perimeter() > s2->perimeter();
+		};
+	}
+
+	// Sort and return new ShapeSorter for sorted vector
+	// of shapes.
+	sort(n_shapes.begin(), n_shapes.end(), sorter);
+	return ShapeSorter(n_shapes);
+}
+
+/*
+ * Filters the vector of shapes by their type (name).
+ *
+ * @param type	The type to filter by.
+ *
+ * @return		A ShapeSorter for the filtered vector of shapes.
+ */
+ShapeSorter ShapeSorter::filterByType(std::string type) const {
+
+	// Create type filter function.
+	auto filter = [&type](const Shape* s) -> bool {
```

Comment:
It might have been good to use an enum class instead of a string in the shape classes. This would make the comparison much more efficient.